File: Plugins/AM_DtaRefinery_PlugIn/clsAnalysisToolRunnerDtaRefinery.vb
Option Strict On

'*********************************************************************************************************
' Written by Matt Monroe for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Copyright 2009, Battelle Memorial Institute
'
'*********************************************************************************************************

Imports AnalysisManagerBase

Public Class clsAnalysisToolRunnerDtaRefinery
    Inherits clsAnalysisToolRunnerBase

    '*********************************************************************************************************
    'Class for running DTA_Refinery analysis
    '*********************************************************************************************************

#Region "Module Variables"
    Protected Const PROGRESS_PCT_DTA_REFINERY_RUNNING As Single = 5
    Protected Const PROGRESS_PCT_PEPTIDEHIT_START As Single = 95
    Protected Const PROGRESS_PCT_PEPTIDEHIT_COMPLETE As Single = 99

    Protected WithEvents CmdRunner As clsRunDosProgram
    '--------------------------------------------------------------------------------------------
    'Future section to monitor DTA_Refinery log file for progress determination
    '--------------------------------------------------------------------------------------------
    'Dim WithEvents m_StatFileWatch As FileSystemWatcher
    'Protected m_XtSetupFile As String = "default_input.xml"
    '--------------------------------------------------------------------------------------------
    'End future section
    '--------------------------------------------------------------------------------------------
#End Region

#Region "Methods"
    ''' <summary>
    ''' Runs DTA_Refinery tool
    ''' </summary>
    ''' <returns>CloseOutType enum indicating success or failure</returns>
    ''' <remarks></remarks>
    Public Overrides Function RunTool() As IJobParams.CloseOutType

        Dim result As IJobParams.CloseOutType
        Dim OrgDBName As String = m_jobParams.GetParam("PeptideSearch", "generatedFastaName")
        Dim LocalOrgDBFolder As String = m_mgrParams.GetParam("orgdbdir")

        'Do the base class stuff
        If Not MyBase.RunTool = IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        ' Store the DTARefinery and X!Tandem version info in the database
		If Not StoreToolVersionInfo() Then
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "Aborting since StoreToolVersionInfo returned false")
			m_message = "Error determining DTA Refinery version"
			Return IJobParams.CloseOutType.CLOSEOUT_FAILED
		End If

        ' Make sure the _DTA.txt file is valid
        If Not ValidateCDTAFile() Then
            Return IJobParams.CloseOutType.CLOSEOUT_NO_DTA_FILES
        End If

        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.INFO, "Running DTA_Refinery")

        CmdRunner = New clsRunDosProgram(m_WorkDir)

        If m_DebugLevel > 4 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "clsAnalysisToolRunnerDtaRefinery.RunTool(): Enter")
        End If

        ' verify that program file exists
        ' DTARefineryLoc will be something like this: "c:\dms_programs\DTARefinery\dta_refinery.exe"
        Dim progLoc As String = m_mgrParams.GetParam("DTARefineryLoc")
        If Not System.IO.File.Exists(progLoc) Then
            If progLoc.Length = 0 Then progLoc = "Parameter 'DTARefineryLoc' not defined for this manager"
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "Cannot find DTA_Refinery program file: " & progLoc)
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        Dim CmdStr As String
        CmdStr = System.IO.Path.Combine(m_WorkDir, m_jobParams.GetParam("DTARefineryXMLFile"))
        CmdStr &= " " & System.IO.Path.Combine(m_WorkDir, m_Dataset & "_dta.txt")
        CmdStr &= " " & System.IO.Path.Combine(LocalOrgDBFolder, OrgDBName)

		If m_DebugLevel >= 1 Then
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, progLoc & " " & CmdStr)
		End If
        If Not CmdRunner.RunProgram(progLoc, CmdStr, "DTARefinery", True) Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogDb, clsLogTools.LogLevels.ERROR, "Error running DTARefinery, job " & m_JobNum)

            ValidateDTARefineryLogFile()

            ' Move the source files and any results to the Failed Job folder
            ' Useful for debugging DTA_Refinery problems
            CopyFailedResultsToArchiveFolder()

            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        'Stop the job timer
        m_StopTime = System.DateTime.UtcNow

        'Add the current job data to the summary file
        If Not UpdateSummaryFile() Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogDb, clsLogTools.LogLevels.WARN, "Error creating summary file, job " & m_JobNum & ", step " & m_jobParams.GetParam("Step"))
        End If

        'Make sure objects are released
        System.Threading.Thread.Sleep(2000)        '2 second delay
        GC.Collect()
        GC.WaitForPendingFinalizers()

        If Not ValidateDTARefineryLogFile() Then
            result = IJobParams.CloseOutType.CLOSEOUT_NO_DATA
        Else
            'Zip the output file
            result = ZipMainOutputFile()
        End If

        If result <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            ' Move the source files and any results to the Failed Job folder
            ' Useful for debugging DTA_Refinery problems
            CopyFailedResultsToArchiveFolder()
            Return result
        End If

        result = MakeResultsFolder()
        If result <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            'TODO: What do we do here?
            Return result
        End If

        result = MoveResultFiles()
        If result <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            'TODO: What do we do here?
            ' Note that MoveResultFiles should have already called clsAnalysisResults.CopyFailedResultsToArchiveFolder
            Return result
        End If

        result = CopyResultsFolderToServer()
        If result <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            'TODO: What do we do here?
            ' Note that MoveResultFiles should have already called clsAnalysisResults.CopyFailedResultsToArchiveFolder
            Return result
        End If

        Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS 'ZipResult

    End Function

    Protected Sub CopyFailedResultsToArchiveFolder()

        Dim result As IJobParams.CloseOutType

        Dim strFailedResultsFolderPath As String = m_mgrParams.GetParam("FailedResultsFolderPath")
        If String.IsNullOrEmpty(strFailedResultsFolderPath) Then strFailedResultsFolderPath = "??Not Defined??"

        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "Processing interrupted; copying results to archive folder: " & strFailedResultsFolderPath)

        ' Bump up the debug level if less than 2
        If m_DebugLevel < 2 Then m_DebugLevel = 2

        ' Try to save whatever files are in the work directory (however, delete the _DTA.txt and _DTA.zip files first)
        Dim strFolderPathToArchive As String
        strFolderPathToArchive = String.Copy(m_WorkDir)

        Try
            System.IO.File.Delete(System.IO.Path.Combine(m_WorkDir, m_Dataset & "_dta.zip"))
            System.IO.File.Delete(System.IO.Path.Combine(m_WorkDir, m_Dataset & "_dta.txt"))
        Catch ex As Exception
            ' Ignore errors here
        End Try

        ' Make the results folder
        result = MakeResultsFolder()
        If result = IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            ' Move the result files into the result folder
            result = MoveResultFiles()
            If result = IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
                ' Move was a success; update strFolderPathToArchive
                strFolderPathToArchive = System.IO.Path.Combine(m_WorkDir, m_ResFolderName)
            End If
        End If

        ' Copy the results folder to the Archive folder
        Dim objAnalysisResults As clsAnalysisResults = New clsAnalysisResults(m_mgrParams, m_jobParams)
        objAnalysisResults.CopyFailedResultsToArchiveFolder(strFolderPathToArchive)


    End Sub

    ''' <summary>
    ''' Stores the tool version info in the database
    ''' </summary>
    ''' <remarks></remarks>
    Protected Function StoreToolVersionInfo() As Boolean

        Dim strToolVersionInfo As String = String.Empty

        If m_DebugLevel >= 2 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "Determining tool version info")
        End If

        ' Store paths to key files in ioToolFiles
        Dim ioToolFiles As New System.Collections.Generic.List(Of System.IO.FileInfo)
        Dim ioDtaRefineryFileInfo As New System.IO.FileInfo(m_mgrParams.GetParam("DTARefineryLoc"))

        If ioDtaRefineryFileInfo.Exists Then
            ioToolFiles.Add(ioDtaRefineryFileInfo)

            Dim strXTandemModuleLoc As String = System.IO.Path.Combine(ioDtaRefineryFileInfo.DirectoryName, "aux_xtandem_module\tandem_5digit_precision.exe")
			ioToolFiles.Add(New System.IO.FileInfo(strXTandemModuleLoc))
		Else
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "DTARefinery not found: " & ioDtaRefineryFileInfo.FullName)
			Return False
		End If

        Try
            Return MyBase.SetStepTaskToolVersion(strToolVersionInfo, ioToolFiles)
        Catch ex As Exception
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "Exception calling SetStepTaskToolVersion: " & ex.Message)
            Return False
        End Try

    End Function

    ''' <summary>
    ''' Parses the _DTARefineryLog.txt file to look for errors
    ''' </summary>
    ''' <returns>True if no errors, false if a problem</returns>
    ''' <remarks></remarks>
    Public Function ValidateDTARefineryLogFile() As Boolean

        Dim ioSourceFile As System.IO.FileInfo
        Dim srSourceFile As System.IO.StreamReader

        Dim strLineIn As String

        Try

            ioSourceFile = New System.IO.FileInfo(System.IO.Path.Combine(m_WorkDir, m_Dataset & "_dta_DtaRefineryLog.txt"))
            If Not ioSourceFile.Exists Then
                m_message = "DtaRefinery Log file not found"
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & " (" & ioSourceFile.Name & ")")
                Return False
            End If

            srSourceFile = New System.IO.StreamReader(New System.IO.FileStream(ioSourceFile.FullName, IO.FileMode.Open, IO.FileAccess.Read, IO.FileShare.Read))

            Do While srSourceFile.Peek > -1
                strLineIn = srSourceFile.ReadLine()

                If strLineIn.StartsWith("number of spectra identified less than 2") Then
                    If srSourceFile.Peek > -1 Then
                        strLineIn = srSourceFile.ReadLine()
                        If strLineIn.StartsWith("stop processing") Then
                            m_message = "X!Tandem identified fewer than 2 peptides; unable to use DTARefinery with this dataset"
                            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
                            Return False
                        End If
                    End If

                    clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "Encountered message 'number of spectra identified less than 2' but did not find 'stop processing' on the next line; DTARefinery likely did not complete properly")

                End If
            Loop

            srSourceFile.Close()

        Catch ex As Exception
            m_message = "Exception in ValidateDTARefineryLogFile"
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & ": " & ex.Message)
            Return False
        End Try

        Return True

    End Function


    ''' <summary>
    ''' Zips concatenated XML output file
    ''' </summary>
    ''' <returns>CloseOutType enum indicating success or failure</returns>
    ''' <remarks></remarks>
    Private Function ZipMainOutputFile() As IJobParams.CloseOutType

        Dim ioWorkDirectory As System.IO.DirectoryInfo
        Dim ioFiles() As System.IO.FileInfo
        Dim ioFile As System.IO.FileInfo
        Dim strFixedDTAFilePath As String

        'Do we want to zip these output files?  Yes, we keep them all
        '* _dta_DtaRefineryLog.txt 
        '* _dta_SETTINGS.xml
        '* _FIXED_dta.txt
        '* _HIST.png
        '* _HIST.txt
        ' * scan number: _scanNum.png
        ' * m/z: _mz.png
        ' * log10 of ion intensity in the ICR/Orbitrap cell: _logTrappedIonInt.png
        ' * total ion current in the ICR/Orbitrap cell: _trappedIonsTIC.png


        'Delete the original DTA files
        Try
            ioWorkDirectory = New System.IO.DirectoryInfo(m_WorkDir)
            ioFiles = ioWorkDirectory.GetFiles("*_dta.*")

            For Each ioFile In ioFiles
                If Not ioFile.Name.ToUpper.EndsWith("_FIXED_dta.txt".ToUpper) Then
                    ioFile.Attributes = ioFile.Attributes And (Not System.IO.FileAttributes.ReadOnly)
                    ioFile.Delete()
                End If
            Next

        Catch Err As Exception
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "clsAnalysisToolRunnerDtaRefinery.ZipMainOutputFile, Error deleting _om.omx file, job " & m_JobNum & Err.Message)
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End Try

        Try
            strFixedDTAFilePath = System.IO.Path.Combine(m_WorkDir, m_Dataset & "_FIXED_dta.txt")
            ioFile = New System.IO.FileInfo(strFixedDTAFilePath)

            If Not ioFile.Exists Then
                Dim Msg As String = "DTARefinery output file not found"
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, Msg & ": " & ioFile.Name)
				m_message = clsGlobal.AppendToComment(m_message, Msg)
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            ioFile.MoveTo(System.IO.Path.Combine(m_WorkDir, m_Dataset & "_dta.txt"))

            Try
                If Not MyBase.ZipFile(ioFile.FullName, True) Then
                    Dim Msg As String = "Error zipping DTARefinery output file"
                    clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, Msg & ": " & ioFile.FullName)
					m_message = clsGlobal.AppendToComment(m_message, Msg)
                    Return IJobParams.CloseOutType.CLOSEOUT_FAILED
                End If

            Catch ex As Exception
                Dim Msg As String = "clsAnalysisToolRunnerDtaRefinery.ZipMainOutputFile, Error zipping DTARefinery output file: " & ex.Message
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, Msg)
				m_message = clsGlobal.AppendToComment(m_message, "Error zipping DTARefinery output file")
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End Try

        Catch ex As Exception
            Dim Msg As String = "clsAnalysisToolRunnerDtaRefinery.ZipMainOutputFile, Error renaming DTARefinery output file: " & ex.Message
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, Msg)
			m_message = clsGlobal.AppendToComment(m_message, "Error renaming DTARefinery output file")
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End Try

        Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS

    End Function

    ''' <summary>
    ''' Event handler for CmdRunner.LoopWaiting event
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub CmdRunner_LoopWaiting() Handles CmdRunner.LoopWaiting
        Static dtLastStatusUpdate As System.DateTime = System.DateTime.UtcNow

        ' Synchronize the stored Debug level with the value stored in the database
        Const MGR_SETTINGS_UPDATE_INTERVAL_SECONDS As Integer = 300
        MyBase.GetCurrentMgrSettingsFromDB(MGR_SETTINGS_UPDATE_INTERVAL_SECONDS)

        'Update the status file (limit the updates to every 5 seconds)
        If System.DateTime.UtcNow.Subtract(dtLastStatusUpdate).TotalSeconds >= 5 Then
            dtLastStatusUpdate = System.DateTime.UtcNow
            m_StatusTools.UpdateAndWrite(IStatusFile.EnumMgrStatus.RUNNING, IStatusFile.EnumTaskStatus.RUNNING, IStatusFile.EnumTaskStatusDetail.RUNNING_TOOL, PROGRESS_PCT_DTA_REFINERY_RUNNING, 0, "", "", "", False)
        End If

    End Sub

    '--------------------------------------------------------------------------------------------
    'Future section to monitor log file for progress determination
    '--------------------------------------------------------------------------------------------
    '	Private Sub StartFileWatcher(ByVal DirToWatch As String, ByVal FileToWatch As String)

    ''Watches the DTA_Refinery status file and reports changes

    ''Setup
    'm_StatFileWatch = New FileSystemWatcher
    'With m_StatFileWatch
    '	.BeginInit()
    '	.Path = DirToWatch
    '	.IncludeSubdirectories = False
    '	.Filter = FileToWatch
    '	.NotifyFilter = NotifyFilters.LastWrite Or NotifyFilters.Size
    '	.EndInit()
    'End With

    ''Start monitoring
    'm_StatFileWatch.EnableRaisingEvents = True

    '	End Sub
    '--------------------------------------------------------------------------------------------
    'End future section
    '--------------------------------------------------------------------------------------------
#End Region

End Class

File: Plugins/AM_DtaRefinery_PlugIn/clsAnalysisToolRunnerDtaRefinery.cs
//*********************************************************************************************************
// Written by Matt Monroe for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Copyright 2009, Battelle Memorial Institute
//
//*********************************************************************************************************

using System;
using System.Collections.Generic;
using System.IO;
using AnalysisManagerBase;
using PRISM;

namespace AnalysisManagerDtaRefineryPlugIn
{
    /// <summary>
    /// Class for running DTA_Refinery analysis
    /// </summary>
    public class clsAnalysisToolRunnerDtaRefinery : clsAnalysisToolRunnerBase
    {
        #region "Module Variables"

        private const float PROGRESS_PCT_DTA_REFINERY_RUNNING = 5;

        private bool mXTandemHasFinished;

        private clsRunDosProgram mCmdRunner;

        #endregion

        #region "Methods"

        /// <summary>
        /// Runs DTA_Refinery tool
        /// </summary>
        /// <returns>CloseOutType enum indicating success or failure</returns>
        public override CloseOutType RunTool()
        {
            CloseOutType result;
            var orgDBName = mJobParams.GetParam("PeptideSearch", "generatedFastaName");
            var localOrgDBFolder = mMgrParams.GetParam("OrgDbDir");

            // Do the base class stuff
            if (base.RunTool() != CloseOutType.CLOSEOUT_SUCCESS)
            {
                return CloseOutType.CLOSEOUT_FAILED;
            }

            if (mDebugLevel > 4)
            {
                LogDebug("clsAnalysisToolRunnerDtaRefinery.RunTool(): Enter");
            }

            // Store the DTARefinery and X!Tandem version info in the database
            if (!StoreToolVersionInfo())
            {
                LogError("Aborting since StoreToolVersionInfo returned false");
                mMessage = "Error determining DTA Refinery version";
                return CloseOutType.CLOSEOUT_FAILED;
            }

            // Make sure the _DTA.txt file is valid
            if (!ValidateCDTAFile())
            {
                return CloseOutType.CLOSEOUT_NO_DTA_FILES;
            }

            if (mDebugLevel >= 2)
            {
                LogMessage("Running DTA_Refinery");
            }

            mCmdRunner = new clsRunDosProgram(mWorkDir, mDebugLevel)
            {
                CreateNoWindow = false,
                EchoOutputToConsole = false,
                CacheStandardOutput = false,
                WriteConsoleOutputToFile = false
            };

            RegisterEvents(mCmdRunner);
            mCmdRunner.LoopWaiting += CmdRunner_LoopWaiting;

            // verify that program file exists
            // DTARefineryLoc will be something like this: "c:\dms_programs\DTARefinery\dta_refinery.exe"
            var progLoc = mMgrParams.GetParam("DTARefineryLoc");
            if (!File.Exists(progLoc))
            {
                if (progLoc.Length == 0)
                    progLoc = "Parameter 'DTARefineryLoc' not defined for this manager";
                LogError("Cannot find DTA_Refinery program file: " + progLoc);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            var arguments =
                Path.Combine(mWorkDir, mJobParams.GetParam("DTARefineryXMLFile")) + " " +
                Path.Combine(mWorkDir, mDatasetName + "_dta.txt") + " " +
                Path.Combine(localOrgDBFolder, orgDBName);

            // Create a batch file to run the command
            // Capture the console output (including output to the error stream) via redirection symbols:
            //    exePath arguments > ConsoleOutputFile.txt 2>&1

            var strBatchFilePath = Path.Combine(mWorkDir, "Run_DTARefinery.bat");
            var strConsoleOutputFileName = "DTARefinery_Console_Output.txt";
            mJobParams.AddResultFileToSkip(Path.GetFileName(strBatchFilePath));

            var strBatchFileCmdLine = progLoc + " " + arguments + " > " + strConsoleOutputFileName + " 2>&1";

            // Create the batch file
            using (var writer = new StreamWriter(new FileStream(strBatchFilePath, FileMode.Create, FileAccess.Write, FileShare.Read)))
            {
                if (mDebugLevel >= 1)
                {
                    LogDebug(strBatchFileCmdLine);
                }

                writer.WriteLine(strBatchFileCmdLine);
            }

            mProgress = PROGRESS_PCT_DTA_REFINERY_RUNNING;
            ResetProgRunnerCpuUsage();
            mXTandemHasFinished = false;

            // Start the program and wait for it to finish
            // However, while it's running, LoopWaiting will get called via events
            var processingSuccess = mCmdRunner.RunProgram(strBatchFilePath, string.Empty, "DTARefinery", true);

            if (!processingSuccess)
            {
                clsGlobal.IdleLoop(0.5);

                // Open DTARefinery_Console_Output.txt and look for the last line with the text "error"
                var fiConsoleOutputFile = new FileInfo(Path.Combine(mWorkDir, strConsoleOutputFileName));
                var consoleOutputErrorMessage = string.Empty;

                if (fiConsoleOutputFile.Exists)
                {
                    using (var consoleOutputReader = new StreamReader(new FileStream(fiConsoleOutputFile.FullName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite)))
                    {
                        while (!consoleOutputReader.EndOfStream)
                        {
                            var dataLine = consoleOutputReader.ReadLine();
                            if (string.IsNullOrWhiteSpace(dataLine))
                            {
                                continue;
                            }

                            if (dataLine.IndexOf("error", StringComparison.InvariantCultureIgnoreCase) >= 0)
                            {
                                consoleOutputErrorMessage = string.Copy(dataLine);
                            }
                        }
                    }
                }

                var msg = "Error running DTARefinery";
                if (!string.IsNullOrWhiteSpace(consoleOutputErrorMessage))
                {
                    msg += ": " + consoleOutputErrorMessage;
                }

                mMessage = string.Empty;
                LogError(msg);

                ValidateDTARefineryLogFile();

                // Move the source files and any results to the Failed Job folder
                // Useful for debugging DTA_Refinery problems
                CopyFailedResultsToArchiveDirectory();

                return CloseOutType.CLOSEOUT_FAILED;
            }

            // Stop the job timer
            mStopTime = DateTime.UtcNow;

            // Add the current job data to the summary file
            UpdateSummaryFile();

            // Make sure objects are released
            clsGlobal.IdleLoop(0.5);
            ProgRunner.GarbageCollectNow();

            if (!ValidateDTARefineryLogFile())
            {
                result = CloseOutType.CLOSEOUT_NO_DATA;
            }
            else
            {
                var oMassErrorExtractor = new clsDtaRefLogMassErrorExtractor(mMgrParams, mWorkDir, mDebugLevel, blnPostResultsToDB: true);
                RegisterEvents(oMassErrorExtractor);

                var intDatasetID = mJobParams.GetJobParameter("DatasetID", 0);

                var blnSuccess = oMassErrorExtractor.ParseDTARefineryLogFile(mDatasetName, intDatasetID, mJob);

                if (!blnSuccess)
                {
                    mMessage = "Error parsing DTA refinery log file to extract mass error stats";
                    LogErrorToDatabase(mMessage + ", job " + mJob);
                }

                // Zip the output file
                result = ZipMainOutputFile();
            }

            if (result != CloseOutType.CLOSEOUT_SUCCESS)
            {
                // Move the source files and any results to the Failed Job folder
                // Useful for debugging DTA_Refinery problems
                CopyFailedResultsToArchiveDirectory();
                return result;
            }

            var success = CopyResultsToTransferDirectory();

            return success ? CloseOutType.CLOSEOUT_SUCCESS : CloseOutType.CLOSEOUT_FAILED;


        }

        /// <summary>
        /// Copy failed results from the working directory to the DMS_FailedResults directory on the local computer
        /// </summary>
        public override void CopyFailedResultsToArchiveDirectory()
        {
            mJobParams.AddResultFileToSkip(Dataset + "_dta.zip");
            mJobParams.AddResultFileToSkip(Dataset + "_dta.txt");

            base.CopyFailedResultsToArchiveDirectory();
        }

        /// <summary>
        /// Parses the _DTARefineryLog.txt file to check for a message regarding X!Tandem being finished
        /// </summary>
        /// <returns>True if finished, false if not</returns>
        /// <remarks></remarks>
        private bool IsXTandemFinished()
        {
            try
            {
                var fiSourceFile = new FileInfo(Path.Combine(mWorkDir, mDatasetName + "_dta_DtaRefineryLog.txt"));
                if (!fiSourceFile.Exists)
                {
                    LogDebug("DTA_Refinery log file not found by IsXTandemFinished: " + fiSourceFile.Name, 10);
                    return false;
                }

                var tmpFilePath = fiSourceFile.FullName + ".tmp";
                fiSourceFile.CopyTo(tmpFilePath, true);
                mJobParams.AddResultFileToSkip(tmpFilePath);
                clsGlobal.IdleLoop(0.1);

                using (var reader = new StreamReader(new FileStream(tmpFilePath, FileMode.Open, FileAccess.Read, FileShare.Read)))
                {
                    while (!reader.EndOfStream)
                    {
                        var dataLine = reader.ReadLine();

                        if (dataLine != null && dataLine.Contains("finished x!tandem"))
                        {
                            LogMessage("X!Tandem has finished searching and now DTA_Refinery is running (parsed " + fiSourceFile.Name + ")");
                            return true;
                        }
                    }
                }

                return false;
            }
            catch (Exception ex)
            {
                LogError("Exception in IsXTandemFinished: " + ex.Message);
                return false;
            }
        }

        /// <summary>
        /// Stores the tool version info in the database
        /// </summary>
        /// <remarks></remarks>
        private bool StoreToolVersionInfo()
        {
            var strToolVersionInfo = string.Empty;

            if (mDebugLevel >= 2)
            {
                LogDebug("Determining tool version info");
            }

            // Store paths to key files in toolFiles
            var toolFiles = new List<FileInfo>();
            var ioDtaRefineryFileInfo = new FileInfo(mMgrParams.GetParam("DTARefineryLoc"));

            if (ioDtaRefineryFileInfo.Exists)
            {
                toolFiles.Add(ioDtaRefineryFileInfo);

                if (ioDtaRefineryFileInfo.DirectoryName == null)
                {
                    LogError("Unable to determine the parent directory of " + ioDtaRefineryFileInfo.FullName);
                }
                else
                {
                    var strXTandemModuleLoc = Path.Combine(ioDtaRefineryFileInfo.DirectoryName, @"aux_xtandem_module\tandem_5digit_precision.exe");
                    toolFiles.Add(new FileInfo(strXTandemModuleLoc));
                }
            }
            else
            {
                LogError("DTARefinery not found: " + ioDtaRefineryFileInfo.FullName);
                return false;
            }

            try
            {
                return SetStepTaskToolVersion(strToolVersionInfo, toolFiles, saveToolVersionTextFile: false);
            }
            catch (Exception ex)
            {
                LogError("Exception calling SetStepTaskToolVersion: " + ex.Message, ex);
                return false;
            }
        }

        /// <summary>
        /// Parses the _DTARefineryLog.txt file to look for errors
        /// </summary>
        /// <returns>True if no errors, false if a problem</returns>
        /// <remarks></remarks>
        private bool ValidateDTARefineryLogFile()
        {
            try
            {
                var fiSourceFile = new FileInfo(Path.Combine(mWorkDir, mDatasetName + "_dta_DtaRefineryLog.txt"));
                if (!fiSourceFile.Exists)
                {
                    mMessage = string.Empty;
                    LogError("DtaRefinery Log file not found (" + fiSourceFile.Name + ")");
                    return false;
                }

                using (var reader = new StreamReader(new FileStream(fiSourceFile.FullName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite)))
                {
                    while (!reader.EndOfStream)
                    {
                        var dataLine = reader.ReadLine();

                        if (dataLine == null || !dataLine.StartsWith("number of spectra identified less than 2", StringComparison.InvariantCultureIgnoreCase))
                            continue;

                        if (!reader.EndOfStream)
                        {
                            dataLine = reader.ReadLine();
                            if (dataLine != null && dataLine.StartsWith("stop processing", StringComparison.InvariantCultureIgnoreCase))
                            {
                                mMessage = string.Empty;
                                LogError("X!Tandem identified fewer than 2 peptides; unable to use DTARefinery with this dataset");
                                return false;
                            }
                        }

                        LogWarning("Encountered message 'number of spectra identified less than 2' but did not find 'stop processing' on the next line; " +
                                   "DTARefinery likely did not complete properly");
                    }
                }
            }
            catch (Exception ex)
            {
                LogError("Exception in ValidateDTARefineryLogFile: " + ex.Message, ex);
                return false;
            }

            return true;
        }

        /// <summary>
        /// Zips concatenated XML output file
        /// </summary>
        /// <returns>CloseOutType enum indicating success or failure</returns>
        /// <remarks></remarks>
        private CloseOutType ZipMainOutputFile()
        {
            // Do we want to zip these output files?  Yes, we keep them all
            // * _dta_DtaRefineryLog.txt
            // * _dta_SETTINGS.xml
            // * _FIXED_dta.txt
            // * _HIST.png
            // * _HIST.txt
            // * scan number: _scanNum.png
            // * m/z: _mz.png
            // * log10 of ion intensity in the ICR/Orbitrap cell: _logTrappedIonInt.png
            // * total ion current in the ICR/Orbitrap cell: _trappedIonsTIC.png

            // Delete the original DTA files
            var targetFile = "??";

            try
            {
                var ioWorkDirectory = new DirectoryInfo(mWorkDir);
                var ioFiles = ioWorkDirectory.GetFiles("*_dta.*");

                foreach (var ioFile in ioFiles)
                {
                    if (!ioFile.Name.EndsWith("_FIXED_dta.txt", StringComparison.InvariantCultureIgnoreCase))
                    {
                        targetFile = ioFile.Name;
                        ioFile.Attributes = ioFile.Attributes & ~FileAttributes.ReadOnly;
                        ioFile.Delete();
                    }
                }
            }
            catch (Exception ex)
            {
                LogError(string.Format("Error deleting file {0}: {1}", targetFile, ex.Message), ex);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            try
            {
                var fixedDTAFilePath = Path.Combine(mWorkDir, mDatasetName + "_FIXED_dta.txt");
                var fixedDtaFile = new FileInfo(fixedDTAFilePath);

                if (!fixedDtaFile.Exists)
                {
                    LogError("DTARefinery output file not found: " + fixedDtaFile.Name);
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                fixedDtaFile.MoveTo(Path.Combine(mWorkDir, mDatasetName + "_dta.txt"));

                try
                {
                    if (!ZipFile(fixedDtaFile.FullName, true))
                    {
                        LogError("Error zipping DTARefinery output file: " + fixedDtaFile.FullName);
                        return CloseOutType.CLOSEOUT_FAILED;
                    }
                }
                catch (Exception ex)
                {
                    LogError("Error zipping DTARefinery output file: " + ex.Message, ex);
                    return CloseOutType.CLOSEOUT_FAILED;
                }
            }
            catch (Exception ex)
            {
                LogError("Error renaming DTARefinery output file: " + ex.Message, ex);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            return CloseOutType.CLOSEOUT_SUCCESS;
        }

        private DateTime mLastCpuUsageCheck = DateTime.MinValue;

        /// <summary>
        /// Event handler for CmdRunner.LoopWaiting event
        /// </summary>
        /// <remarks></remarks>
        private void CmdRunner_LoopWaiting()
        {
            const string DTA_REFINERY_PROCESS_NAME = "dta_refinery";
            const string XTANDEM_PROCESS_NAME = "tandem_5digit_precision";
            const int SECONDS_BETWEEN_UPDATE = 30;

            UpdateStatusFile();

            // Push a new core usage value into the queue every 30 seconds
            if (DateTime.UtcNow.Subtract(mLastCpuUsageCheck).TotalSeconds >= SECONDS_BETWEEN_UPDATE)
            {
                mLastCpuUsageCheck = DateTime.UtcNow;

                if (!mXTandemHasFinished)
                {
                    mXTandemHasFinished = IsXTandemFinished();
                }

                if (mXTandemHasFinished)
                {
                    // Determine the CPU usage of DTA_Refinery
                    UpdateCpuUsageByProcessName(DTA_REFINERY_PROCESS_NAME, SECONDS_BETWEEN_UPDATE, mCmdRunner.ProcessID);
                }
                else
                {
                    // Determine the CPU usage of X!Tandem
                    UpdateCpuUsageByProcessName(XTANDEM_PROCESS_NAME, SECONDS_BETWEEN_UPDATE, mCmdRunner.ProcessID);
                }

                LogProgress("DtaRefinery");
            }
        }

        #endregion
    }
}
